File: ports/java/src/core/global/BcsPointer.java
package bcs.core.global;

// Java uses GC, so these are semantic wrappers mapping the C++ ownership concepts.
public class BcsPointer {
<<<<<<< HEAD

=======

>>>>>>> origin/master
    public static class BcsSharedPointer<T> {
        private T value;

        public BcsSharedPointer(T value) {
            this.value = value;
        }

        public T get() { return value; }
    }

    public static class BcsUniquePointer<T extends AutoCloseable> implements AutoCloseable {
        private T value;

        public BcsUniquePointer(T value) {
            this.value = value;
        }

        public T get() { return value; }

        @Override
        public void close() throws Exception {
            if (value != null) {
                value.close();
                value = null;
            }
        }
    }
}

File: ports/csharp/src/core/global/BcsPointer.cs
using System;

namespace BCS.Core.Global {
<<<<<<< HEAD
    // C# uses a GC, so BcsSharedPointer/BcsUniquePointer are semantic wrappers
=======
    // C# uses a GC, so BcsSharedPointer/BcsUniquePointer are semantic wrappers
>>>>>>> origin/master
    // or just direct object references. We provide a lightweight generic wrapper for parity.
    public class BcsSharedPointer<T> where T : class {
        public T Value { get; private set; }

        public BcsSharedPointer(T value) {
            Value = value;
        }
    }

    public class BcsUniquePointer<T> : IDisposable where T : IDisposable {
        public T Value { get; private set; }

        public BcsUniquePointer(T value) {
            Value = value;
        }

        public void Dispose() {
            if (Value != null) {
                Value.Dispose();
                Value = default(T);
            }
        }
    }
}
